n3fit/src/n3fit/layers/msr_normalization.py

```diff
@@ -51,30 +83,35 @@ def call(self, pdf_integrated, photon_integral):
 
         Parameters
         ----------
-        pdf_integrated: (Tensor(1,None,14))
+        pdf_integrated: (Tensor(1, 14))
             the integrated PDF
-        photon_integral: (Tensor(1)):
-            the integrated photon, not included in PDF
+        photon_integral: (Tensor(1, 1))
+            the integrated photon PDF
 
         Returns
         -------
         normalization_factor: Tensor(14)
             The normalization factors per flavour.
         """
-        y = op.flatten(pdf_integrated)
+        y = pdf_integrated[0]  # get rid of the batch dimension
+        photon_integral = photon_integral[0]  # get rid of the batch dimension
         norm_constants = []
 
         if self._msr_enabled:
-            n_ag = [(1.0 - y[GLUON_IDX[0][0] - 1] - photon_integral[0]) / y[GLUON_IDX[0][0]]] * len(
-                GLUON_IDX
-            )
-            norm_constants += n_ag
+            norm_constants += [
+                op.batchit(1.0 - y[IDX['sigma']] - photon_integral[0], batch_dimension=0)
+            ]
 
         if self._vsr_enabled:
-            n_av = [3.0 / y[V_IDX[0][0]]] * len(V_IDX)
-            n_av3 = [1.0 / y[V3_IDX[0][0]]] * len(V3_IDX)
-            n_av8 = [3.0 / y[V8_IDX[0][0]]] * len(V8_IDX)
-            n_av15 = [3.0 / y[V15_IDX[0][0]]] * len(V15_IDX)
-            norm_constants += n_av + n_av3 + n_av8 + n_av15
+            norm_constants += [self.vsr_factors]
+        norm_constants = op.concatenate(norm_constants, axis=0)
+
+        divisors = op.gather(y, self.divisor_indices, axis=0)
+        norm_constants = norm_constants / divisors
+
+        # Fill in the rest of the flavours with 1
+        norm_constants = op.scatter_to_one(
+            norm_constants, indices=self.indices, output_shape=y.shape
+        )
 
-        return self._out_scatter(norm_constants)
+        return norm_constants
```

Comment:
Please don't use the same name for all these variables, specially the last one where things have been reorganized